### PR TITLE
improve memory footprint of daemonset simulate

### PR DESCRIPTION
**Pull request by mfojtik**
**What this PR does / why we need it**:

This is an alternative for https://github.com/kubernetes/kubernetes/pull/64915 (it might be not needed if that PR will merge)

During memory profiling of OpenShift, we noticed a significant amount of object allocations done by `IsControlledBy()`. @sttts found that the `GetObjectReferences()` method is doing a deep-copy of the object references. 

screen shot 2018-06-08 at 14 22 59

This PR simplify the `IsControlledBy()` to just iterate over the ownerRefs, without copying them. 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by mfojtik** [comment]
@sttts fixed logic error in for ;-)

File: pkg/controller/daemon/daemon_controller.go [hunk 1]
```diff
@@ -1259,6 +1259,18 @@ func (dsc *DaemonSetsController) simulate(newPod *v1.Pod, node *v1.Node, ds *app
 		})
 	}
 
+	// This is memory efficient copy of metav1.IsControlledBy where we don't deep copy all references.
```
Comment (sttts) [inline pkg/controller/daemon/daemon_controller.go]: not sure we really need this comment in the code. It fits into the commit message. There is nothing wrong or special to access p.ObjectMeta.OwnerReferences directly without an accessor. This comment suggests it is uncommon.

File: pkg/controller/daemon/daemon_controller.go [hunk 2]
```diff
@@ -1274,7 +1286,7 @@ func (dsc *DaemonSetsController) simulate(newPod *v1.Pod, node *v1.Node, ds *app
 		}
 		// ignore pods that belong to the daemonset when taking into account whether
 		// a daemonset should bind to a node.
-		if metav1.IsControlledBy(pod, ds) {
+		if isControlledByDaemonSet(pod) {
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/controller/daemon/daemon_controller.go]: nit: isControlledByThisDaemonSet

File: pkg/controller/daemon/daemon_controller.go [hunk 3]
```diff
@@ -1259,6 +1259,18 @@ func (dsc *DaemonSetsController) simulate(newPod *v1.Pod, node *v1.Node, ds *app
 		})
 	}
 
+	// This is memory efficient copy of metav1.IsControlledBy where we don't deep copy all references.
+	// This should reduce memory allocations as we call this for every single pod on every node for every daemonset
+	// that is not owned by this daemonset.
+	isControlledByDaemonSet := func(p *v1.Pod) bool {
```
Comment (sttts) [inline pkg/controller/daemon/daemon_controller.go]: make it a helper func at the bottom of the file and pass uid in as well.

File: pkg/controller/daemon/daemon_controller.go (hunk 1 above)
Comment (mfojtik) [reply pkg/controller/daemon/daemon_controller.go]: fixed

File: pkg/controller/daemon/daemon_controller.go (hunk 3 above)
Comment (mfojtik) [reply pkg/controller/daemon/daemon_controller.go]: fixed.

**Comment by sttts** [comment]
bazel?

**Comment by mfojtik** [comment]
@sttts TODO added.

**Comment by mikedanese** [comment]
can you add a benchmark?

/approve

File: pkg/controller/daemon/daemon_controller.go [hunk 4]
```diff
@@ -1274,7 +1275,8 @@ func (dsc *DaemonSetsController) simulate(newPod *v1.Pod, node *v1.Node, ds *app
 		}
 		// ignore pods that belong to the daemonset when taking into account whether
 		// a daemonset should bind to a node.
-		if metav1.IsControlledBy(pod, ds) {
+		// TODO: replace this with metav1.IsControlledBy() in 1.12
+		if isControlledByDaemonSet(pod, ds.GetUID()) {
```
Comment (mikedanese) [inline pkg/controller/daemon/daemon_controller.go]: Can you add a benchmark so we can measure the performance when we revert this change?

Comment (sttts) [reply pkg/controller/daemon/daemon_controller.go]: A benchmark of the func? After https://github.com/kubernetes/kubernetes/pull/64915 merged these two are basically the same.

**Comment by jberkus** [comment]
bumping priority so it can merge

/priority critical-urgent
/remove-priority important-soon

**Comment by mdianjun** [comment]
Does this issue lead controller-manager memory usage to linearly increase?
